File: frontend/lib/api/quickchat.ts
/**
 * API Client Functions for Quick Chat
 * 
 * Handles communication with the backend API for:
 * - Quick chat session initiation
 * - Sending messages in quick chat
 * - Direct topic search
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

/**
 * Search result from quick chat topic search
 */
export interface QuickChatSearchResult {
  course_id: string
  course_title: string
  course_color: string | null
  material_id: string
  material_name: string
  page_number: number
  summary: string
  key_terms: string[]
  rank: number
}

/**
 * Response from quick chat topic search
 */
export interface QuickChatSearchResponse {
  found: boolean
  message: string
  results: QuickChatSearchResult[]
}

/**
 * SSE event types from quick chat
 */
export type QuickChatEventType = 
  | 'message' 
  | 'delta' 
  | 'tool_call' 
  | 'search_results' 
  | 'session' 
  | 'stats'
  | 'navigate' 
  | 'open_material'
  | 'pending_navigation'
  | 'error'

/**
 * SSE event from quick chat stream
 */
export interface QuickChatEvent {
  type: QuickChatEventType
  // Message events
  role?: 'assistant' | 'user'
  content?: string
  delta?: string
  // Tool call events
  tool?: string
  // Search results events
  results?: QuickChatSearchResult[]
  // Session events
  thread_id?: string
  mode?: 'discovery' | 'tutoring'
  courses_count?: number
  materials_count?: number
  // Navigate/open_material events
  course_id?: string
  course_title?: string
  material_id?: string
  material_name?: string
  page_number?: number
  // Error events
  error?: string
}

/**
 * Response from quick chat initiate
 */
export interface QuickChatInitiateResponse {
  greeting: string
  thread_id: string
  mode: 'discovery' | 'tutoring'
}

/**
 * Initiate a quick chat session
 * 
 * Returns instantly with greeting and thread_id (no SSE streaming).
 * 
 * @param userId User ID
 * @returns Initiate response with greeting and thread_id
 */
export async function initiateQuickChat(
  userId: string
): Promise<QuickChatInitiateResponse> {
  const url = `${API_URL}/api/quickchat/initiate`
  
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ user_id: userId }),
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Request failed' }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }

  return response.json()
}

/**
 * Pre-warm the quick chat agent for faster first response.
 * Call this after receiving thread_id from initiate, while user is typing.
 * 
 * @param userId User ID
 * @param threadId Thread ID from initiate response
 * @returns Warmup status
 */
export async function warmupQuickChat(
  userId: string,
  threadId: string
): Promise<{ status: string; thread_id?: string; error?: string }> {
  const url = `${API_URL}/api/quickchat/warmup`
  
  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ user_id: userId, thread_id: threadId }),
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Request failed' }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }

  return response.json()
}

/**
 * Send a message in quick chat
 * 
 * @param userId User ID
 * @param message User message
 * @param threadId Thread ID from initiate (for pre-warmed agent)
 * @param context Optional tutoring context
 * @param onEvent Callback for SSE events
 * @returns Cleanup function
 */
export async function sendQuickChatMessage(
  userId: string,
  message: string,
  threadId: string | null,
  context: {
    materialId?: string
    pageNumber?: number
    courseId?: string
  } | null,
  onEvent: (event: QuickChatEvent) => void
): Promise<() => void> {
  const url = `${API_URL}/api/quickchat/message`
  
  const body: Record<string, unknown> = {
    user_id: userId,
    message,
  }
  
  if (threadId) {
    body.thread_id = threadId
  }
  
  if (context?.materialId) {
    body.material_id = context.materialId
    body.page_number = context.pageNumber
    body.course_id = context.courseId
  }

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Request failed' }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }

  const reader = response.body?.getReader()
  const decoder = new TextDecoder()
  let buffer = ''
  let isAborted = false

  const processStream = async () => {
    if (!reader) return

    try {
      while (!isAborted) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const data = line.slice(6).trim()
            if (data === '[DONE]') {
              return
            }
            try {
              const event = JSON.parse(data) as QuickChatEvent
              onEvent(event)
            } catch (e) {
              console.warn('Failed to parse SSE event:', data)
            }
          }
        }
      }
    } catch (error) {
      if (!isAborted) {
        console.error('Error reading stream:', error)
        onEvent({ type: 'error', error: String(error) })
      }
    }
  }

  processStream()

  return () => {
    isAborted = true
    reader?.cancel()
  }
}

/**
 * Response from getMaterialInfo
 */
export interface MaterialInfoResponse {
  material_id: string
  file_name: string
  pdf_url: string
  page_count: number
}

/**
 * Get material info (PDF URL, page count) for inline viewing
 * 
 * @param materialId Material ID
 * @returns Material info including PDF URL
 */
export async function getMaterialInfo(
  materialId: string
): Promise<MaterialInfoResponse> {
  const url = `${API_URL}/api/quickchat/material/${materialId}`
  
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Request failed' }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }

  return response.json()
}

/**
 * Direct topic search without going through the agent
 * 
 * @param userId User ID
 * @param query Search query
 * @param language Language for search ('de', 'en', or 'auto')
 * @param limit Maximum number of results
 * @returns Search response
 */
export async function searchQuickChatTopics(
  userId: string,
  query: string,
  language: string = 'auto',
  limit: number = 10
): Promise<QuickChatSearchResponse> {
  const params = new URLSearchParams({
    user_id: userId,
    query,
    language,
    limit: String(limit),
  })
  
  const url = `${API_URL}/api/quickchat/search?${params}`
  
  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Request failed' }))
    throw new Error(error.detail || `HTTP ${response.status}`)
  }

  return response.json()
}

File: backend/app/api/routers/quickchat.py
import logging

from fastapi import APIRouter, Body, HTTPException, Path, Query
from fastapi.responses import StreamingResponse

from app.adapters.supabase.client import get_supabase_client
from app.core.adapters import get_course, get_page_analysis
from app.models.schemas import (
    QuickChatInitiateRequest,
    QuickChatMessageRequest,
    QuickChatSearchResponse,
    QuickChatSearchResult,
    QuickChatWarmupRequest,
)
from app.services.chat_service import get_chat_service

logger = logging.getLogger(__name__)

router = APIRouter()


def _streaming_headers() -> dict[str, str]:
    return {
        "Cache-Control": "no-cache",
        "Connection": "keep-alive",
        "X-Accel-Buffering": "no",
    }


@router.post("/quickchat/initiate")
async def initiate_quickchat(request: QuickChatInitiateRequest = Body(...)) -> dict:
    return get_chat_service().create_quickchat_session(request.user_id)


@router.post("/quickchat/warmup")
async def warmup_quickchat(request: QuickChatWarmupRequest = Body(...)) -> dict:
    # user_id is intentionally accepted for API parity and future authorization hooks.
    _ = request.user_id
    return await get_chat_service().warmup_quickchat_agent(request.thread_id)


@router.post("/quickchat/message")
async def send_quickchat_message(
    request: QuickChatMessageRequest = Body(...),
) -> StreamingResponse:
    try:
        if not get_course().validate_user_exists(request.user_id):
            raise HTTPException(status_code=404, detail="User not found.")

        stream = get_chat_service().stream_quickchat_message(
            user_id=request.user_id,
            message=request.message,
            thread_id=request.thread_id,
            material_id=request.material_id,
            page_number=request.page_number,
            course_id=request.course_id,
        )

        return StreamingResponse(
            stream,
            media_type="text/event-stream",
            headers=_streaming_headers(),
        )

    except HTTPException:
        raise
    except Exception as exc:
        logger.error("Error sending quick chat message: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to send message: {exc}")


@router.get("/quickchat/search", response_model=QuickChatSearchResponse)
async def search_quickchat_topics(
    user_id: str = Query(...),
    query: str = Query(...),
    language: str = Query("auto"),
    limit: int = Query(10, ge=1, le=50),
) -> QuickChatSearchResponse:
    try:
        if not get_course().validate_user_exists(user_id):
            raise HTTPException(status_code=404, detail="User not found")

        results = get_page_analysis().search(
            user_id=user_id, query=query, language=language, limit=limit
        )

        if not results:
            return QuickChatSearchResponse(
                found=False,
                message=f"Keine Ergebnisse fuer '{query}'.",
                results=[],
            )

        search_results = [
            QuickChatSearchResult(
                course_id=result.get("course_id", ""),
                course_title=result.get("course_title", ""),
                course_color=result.get("course_color"),
                material_id=result.get("material_id", ""),
                material_name=result.get("material_name", ""),
                page_number=result.get("page_number", 0),
                summary=result.get("summary", ""),
                key_terms=result.get("key_terms", []),
                rank=float(result.get("rank", 0)),
            )
            for result in results
        ]
        return QuickChatSearchResponse(
            found=True,
            message=f"{len(search_results)} Ergebnis(se) gefunden.",
            results=search_results,
        )
    except HTTPException:
        raise
    except Exception as exc:
        logger.error("Error searching quickchat topics: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to search topics: {exc}")


@router.get("/quickchat/material/{material_id}")
async def get_quickchat_material_info(material_id: str = Path(...)) -> dict:
    try:
        client = get_supabase_client()
        material_response = (
            client.table("course_materials")
            .select("id, file_name, file_path, page_count, processing_status")
            .eq("id", material_id)
            .single()
            .execute()
        )

        if not material_response.data:
            raise HTTPException(status_code=404, detail="Material not found")

        material = material_response.data
        if material.get("processing_status") != "completed":
            raise HTTPException(status_code=400, detail="Material is still processing")

        file_path = material.get("file_path")
        if not file_path:
            raise HTTPException(status_code=400, detail="Material has no file path")

        signed_url_response = client.storage.from_("course_materials").create_signed_url(
            file_path, 3600
        )
        if not signed_url_response or not signed_url_response.get("signedURL"):
            raise HTTPException(status_code=500, detail="Failed to create signed URL")

        return {
            "material_id": material.get("id"),
            "file_name": material.get("file_name"),
            "pdf_url": signed_url_response.get("signedURL"),
            "page_count": material.get("page_count", 0),
        }

    except HTTPException:
        raise
    except Exception as exc:
        logger.error("Error getting quickchat material info: %s", exc, exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to get material info: {exc}")
